viewing_party/party.py

```diff
@@ -1,23 +1,274 @@
+
+    return friends_unique_movies_list """
+
+# version 2
+
+
+def get_friends_unique_watched(user_data):
+    """
+    Returns a list of movies that at least one of the user's friends has watched, but the user has not.
+
+    Args:
+        user_data (dict): A dictionary containing the user's watched movies and friends' watched movies.
+
+    Returns:
+        list: A list of movie dictionaries that friends have watched but the user has not.
+
+    Written by: Mariya Mokrynska
+    """
+    friends_watched_titles = get_friends_movies(user_data)
+    user_watched_titles = get_user_movies(user_data)
+
+    # Get the set difference: movies that friends have watched but the user has not
+    unique_titles = friends_watched_titles - user_watched_titles
+
+    unique_movies = []
+    added_titles = set()
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] in unique_titles and movie["title"] not in added_titles:
+                unique_movies.append(movie)
+                added_titles.add(movie["title"])
+
+    return unique_movies
+
 
-        
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+
+def get_available_recs(user_data):
+    """
+    Recommends movies the user hasn't watched but their friends have, 
+    available on the user's subscribed streaming services.
+
+    Args:
+    user_data (dict): Contains 'subscriptions' (list of str) and 'friends' (list of dicts with 'watched' movies).
+
+    Returns:
+    list of dict: Recommended movies that the user hasn't watched, are watched by friends, 
+                and are available on the user's subscriptions.
+
+    Written by: Mariya Mokrynska
+    """
+    recommended_movies_list = []
+    user_watched_titles = get_user_movies(user_data)
+    friends_unique_movies_list = get_friends_unique_watched(user_data)
+
+    for friend_movie in friends_unique_movies_list:
+        if friend_movie["title"] not in user_watched_titles and friend_movie["host"] in user_data["subscriptions"]:
+            recommended_movies_list.append(friend_movie)
+
+    return recommended_movies_list
+
+
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
+def get_new_rec_by_genre(user_data):
+    most_watched_genre = get_most_watched_genre(user_data)
+
+    if not most_watched_genre:
+        return []
+
+    rec_movies = []
+    watched_titles = {movie["title"] for movie in user_data["watched"]}
+    seen_movies = set()
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if (
+                movie["genre"] == most_watched_genre and
+                movie["title"] not in watched_titles and
+                movie["title"] not in seen_movies
+            ):
+                rec_movies.append(movie)
+                seen_movies.add(movie["title"])

```

Comment:
This function can be broken down into two sub-problems. One, find the movies that friends have watched and not the user. Two, of those movies select the ones that have the genre of the user's most watched. What function have we written to take care of the first problem?